### PR TITLE
behind the scenes docs

### PR DESCRIPTION
**Pull request by Moocar**
I found myself having to relearn how the guts of Gatsby work every time I make a change. So I created some docs to stop myself having to learn from scratch each time. 

I'm sure some of it could be moved into other parts of the docs (e.g `___NODE` and `frontmatter___date`). But most isn't important for regular users. Just those looking to learn about how Gatsby works behind the scenes.

I've linked in quite a few places to lines of code on Github. I'm 50/50 on this approach. It's great to link from high level descriptions of the code straight to the code itself, but I'm worried these links will become stale. Open to suggestions here.

Lots more to document, but it's a start. Hopefully others knowledgeable with the code base can add to it over time.

I've also used the new [gatsby-remark-graphviz](https://www.npmjs.com/package/gatsby-remark-graphviz) to display graphs, which has the benefit of being able to source control graphs via a language vs an image drawing application.

### CODE REVIEW

**Comment by Moocar** [comment]
Closing this PR as I've updated the docs significantly. Will reopen when they're pushed.

File: docs/docs/schema-generation-behind-the-scenes.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+During the schema generation phase, we must generate what's called a `ProcessedNodeType` in Gatsby. This is a simple structure that builds on top of a [graphql-js GraphQLObjectType](https://graphql.org/graphql-js/type/#graphqlobjecttype). Our goal in the below steps is to infer and construct this object for each unique node type in redux.
+
+The flow is summarized by the below graph. It shows the intermediate transformations that are performed by code in the Gatsby [schema folder](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby/src/schema), finally resulting in the [ProcessedNoteType](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/schema/build-node-types.js#L182). It uses the example of building a `File` GraphQL type.
+
+```dot
+digraph graphname {
+  "pluginFields" [ label = "custom plugin fields\l{\l    publicURL: {\l        type: GraphQLString,\l        resolve(file, a, c) { ... }\l    }\l}\l ", shape = box ];
+  "typeNodes" [ label = "all redux nodes of type\le.g internal.type === `File`", shape = "box" ];
+  "exampleValue" [ label = "exampleValue\l{\l    relativePath: `bogs/my-blog.md`,\l    accessTime: 8292387234\l}\l ", shape = "box" ];
```
[the remark this answers is not in the document]
Comment (Moocar) [reply docs/docs/schema-generation-behind-the-scenes.md]: My favorite typo

File: docs/docs/schema-gql-type.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+```graphql
+query {
+  file( relativePath { eq: "blog/my-blog.md" } ) {
+    childMarkdownRemark { html }
+  }
+}
+```
+
+To resolve `file.childMarkdownRemark`, we filter over all the node's (that we're resolving) `children`, until we find one of type `markdownRemark`, which is returned. Remember that in redux, `children` is a collection of IDs. So as part of this, we lookup the node by ID in redux too.
```
[the remark this answers is not in the document]
Comment (Moocar) [reply docs/docs/schema-gql-type.md]: I think this one is correct as is, but the brackets between `node's` and `children` confuse it. I'll rearrange the sentence.

**Comment by Moocar** [comment]
@m-allanson Sounds good! Yeah, reviewing 1300+ lines of documentation is no easy feat. Thanks for the initial pass. I agree we should just merge it and fix typos/clarifications etc on the fly.

**Comment by fk** [comment]
🎉 ❤️

**Pull request merged**